src/main/kotlin/org/rust/lang/core/completion/RsLambdaExprCompletionProvider.kt

```diff
@@ -0,0 +1,112 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.core.completion
+
+import com.intellij.codeInsight.completion.CompletionParameters
+import com.intellij.codeInsight.completion.CompletionResultSet
+import com.intellij.codeInsight.completion.InsertionContext
+import com.intellij.codeInsight.lookup.LookupElementBuilder
+import com.intellij.codeInsight.template.TemplateResultListener
+import com.intellij.openapiext.isUnitTestMode
+import com.intellij.patterns.ElementPattern
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiErrorElement
+import com.intellij.psi.util.descendantsOfType
+import com.intellij.psi.util.parentOfType
+import com.intellij.util.ProcessingContext
+import org.rust.ide.utils.template.buildAndRunTemplate
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.RsElementTypes.OR
+import org.rust.lang.core.psi.ext.RsElement
+import org.rust.lang.core.psi.ext.elementType
+import org.rust.lang.core.psi.ext.leftSiblings
+import org.rust.lang.core.psi.ext.startOffset
+import org.rust.lang.core.psiElement
+import org.rust.lang.core.types.BoundElement
+import org.rust.lang.core.types.expectedType
+import org.rust.lang.core.types.implLookup
+import org.rust.lang.core.types.ty.*
+import org.rust.openapiext.createSmartPointer
+
+object RsLambdaExprCompletionProvider : RsCompletionProvider() {
+
+    override val elementPattern: ElementPattern<PsiElement>
+        get() = psiElement<PsiElement>()
+            .withSuperParent(2, psiElement<RsPathExpr>())
+
+    override fun addCompletions(
+        parameters: CompletionParameters,
+        context: ProcessingContext,
+        result: CompletionResultSet
+    ) {
+        val element = parameters.position.safeGetOriginalOrSelf()
+        val expr = element.parentOfType<RsExpr>() ?: return
+
+        val paramCount = getParameterCount(expr) ?: return
+        val params = (0 until paramCount).joinToString(", ") { "p${it}" }
+
+        val start = if (expr.leftSiblings.filter { it !is PsiErrorElement }.firstOrNull()?.elementType == OR) {
+            ""
+        } else {
+            "|"
+        }
+        val text = "$start$params| {}"
```

Comment:
I'd not say we should not complete `{}` here. A user can type it if needed

Reply:
I included the body in the template builder - maybe that's even better?
![complete-lambda](https://user-images.githubusercontent.com/4539057/125441813-a9dffdf7-6f41-44f0-b26f-c1c690801e0b.gif)

Reply:
:+1: but let's move the caret between `{}` if a user decides to leave them as is